### PR TITLE
Refactored PaginatorHelper::numbers() in smaller pieces of code.

### PR DESCRIPTION
**Pull request by burzum**

### CODE REVIEW

File: src/View/Helper/PaginatorHelper.php [hunk 1]
```diff
@@ -635,96 +635,180 @@ public function numbers(array $options = [])
...
+        }
+        return $out;
+    }
+
+    /**
+     * Generates the last number for the paginator numbers() method.
+     *
+     * @param StringTemplate $ellipsis
+     * @param array $params Params from the numbers() method.
+     * @param integer $end End number.
```
Comment (dereuromark) [inline src/View/Helper/PaginatorHelper.php]: int

**Comment by dereuromark** [comment]
Looks a lot better with those separate methods I must say.

**Comment by burzum** [comment]
Thanks, but maybe you have some more ideas, it's better but I'm not 100% happy with it as it is.

File: src/View/Helper/PaginatorHelper.php (hunk 1 above)
Comment (burzum) [inline src/View/Helper/PaginatorHelper.php]: OK, thanks.

**Comment by markstory** [comment]
There are a few phpcs errors.

File: src/View/Helper/PaginatorHelper.php [hunk 2]
```diff
@@ -635,96 +635,180 @@ public function numbers(array $options = [])
...
+     */
+    protected function _modulusNumbers($templater, $params, $options)
+    {
         $out = '';
         $ellipsis = $templater->format('ellipsis', []);
 
-        if ($options['modulus'] && $params['pageCount'] > $options['modulus']) {
-            $half = (int)($options['modulus'] / 2);
-            $end = $params['page'] + $half;
+        extract($this->_getNumbersStartAndEnd($params, $options));
```
Comment (markstory) [inline src/View/Helper/PaginatorHelper.php]: Extract has been nothing but trouble in the past. Either reference the keys in the array or make the required local variables.

File: src/View/Helper/PaginatorHelper.php [hunk 3]
```diff
@@ -635,96 +635,180 @@ public function numbers(array $options = [])
...
-                    'url' => $this->generateUrl(['page' => $i], $options['model']),
-                ];
-                $out .= $templater->format('number', $vars);
+        $out .= $options['after'];
+        $out .= $this->_lastNumber($ellipsis, $params, $end, $options);
+        return $out;
+    }
+
+    /**
+     * Generates the last number for the paginator numbers() method.
```
Comment (markstory) [inline src/View/Helper/PaginatorHelper.php]: Copy paste error.

File: src/View/Helper/PaginatorHelper.php [hunk 4]
```diff
@@ -635,96 +635,180 @@ public function numbers(array $options = [])
...
+            $start = 1;
+            $end = $params['page'] + ($options['modulus'] - $params['page']) + 1;
+        }
+        return compact('start', 'end');
+    }
+
+    /**
+     * Formats a number for the paginator number output.
+     *
+     * @param StringTemplate $templater
```
Comment (ADmad) [inline src/View/Helper/PaginatorHelper.php]: `\Cake\View\StringTemplate`

File: src/View/Helper/PaginatorHelper.php [hunk 5]
```diff
@@ -635,96 +635,180 @@ public function numbers(array $options = [])
...
+            'text' => $options['text'],
+            'url' => $this->generateUrl(['page' => $options['page']], $options['model']),
+        ];
+        return $templater->format('number', $vars);
+    }
+
+    /**
+     * Generates the numbers for the paginator numbers() method.
+     *
+     * @param StringTemplate $templater
```
Comment (ADmad) [inline src/View/Helper/PaginatorHelper.php]: `\Cake\View\StringTemplate`

File: src/View/Helper/PaginatorHelper.php [hunk 6]
```diff
@@ -635,96 +635,180 @@ public function numbers(array $options = [])
...
+    /**
+     * Generates the last number for the paginator numbers() method.
+     *
+     * @param StringTemplate $ellipsis
+     * @param array $params Params from the numbers() method.
+     * @param int $start Start number.
+     * @param array $options Options from the numbers() method.
+     * @return string Markup output.
+     */
+    protected function _firstNumber($ellipsis, $params, $start, $options) {
```
Comment (thaJeztah) [inline src/View/Helper/PaginatorHelper.php]: PSR; Curly bracket should be on next line (sorry, I know you don't like it :))

File: src/View/Helper/PaginatorHelper.php [hunk 7]
```diff
@@ -635,96 +635,180 @@ public function numbers(array $options = [])
...
+    /**
+     * Generates the last number for the paginator numbers() method.
+     *
+     * @param StringTemplate $ellipsis
+     * @param array $params Params from the numbers() method.
+     * @param int $end End number.
+     * @param array $options Options from the numbers() method.
+     * @return string Markup output.
+     */
+    protected function _lastNumber($ellipsis, $params, $end, $options) {
```
Comment (thaJeztah) [inline src/View/Helper/PaginatorHelper.php]: same here

File: src/View/Helper/PaginatorHelper.php [hunk 8]
```diff
@@ -635,96 +635,180 @@ public function numbers(array $options = [])
...
-            if ($end != $params['page']) {
+    /**
+     * Generates the numbers for the paginator numbers() method.
+     *
+     * @param StringTemplate $templater
+     * @param array $params Params from the numbers() method.
+     * @param array $options Options from the numbers() method.
+     * @return string Markup output.
+     */
+    protected function _numbers($templater, $params, $options) {
```
Comment (thaJeztah) [inline src/View/Helper/PaginatorHelper.php]: and here

**Comment by lorenzo** [comment]
@burzum poke

**Comment by burzum** [comment]
@lorenzo Yes, OK, I'll see if I can get it done tonight.

File: src/View/Helper/PaginatorHelper.php (hunk 3 above)
Comment (burzum) [inline src/View/Helper/PaginatorHelper.php]: Sorry, I don't get it, what exactly?

File: src/View/Helper/PaginatorHelper.php (hunk 2 above)
Comment (burzum) [inline src/View/Helper/PaginatorHelper.php]: It never caused me trouble, would you mind to explain what kind of badness can happen with extract()? Thank you.

File: src/View/Helper/PaginatorHelper.php (hunk 3 above)
Comment (dereuromark) [inline src/View/Helper/PaginatorHelper.php]: "_firstNumber"

**Comment by markstory** [comment]
Other than @AD7six 's comment this looks good to me.

**Comment by burzum** [comment]
I've changed what @AD7six suggested, thank you.

**Pull request merged**